File: leetcode_C++/p207.cpp
/*
@Author: mrxhar
@Date:   2023-09-09 08:51:37
@Email:  1337904145@qq.com
@Blog:   https://leetcode.cn/problems/course-schedule/?envType=daily-question&envId=2023-09-09
*/

#include <vector>
using std::vector;
#include <queue>
using std::queue;

// class Solution {
// private:
//     vector<vector<int>> graph;
//     vector<int> inDegree;

// public:
//     bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
//         // 需要从一个没有入度得点出发，遍历所有能够到达得点
//         // 建图，有向图
//         // 入度表
//         graph.resize(numCourses);
//         inDegree.resize(numCourses);
//         for (const auto& info : prerequisites) {
//             graph[info[1]].push_back(info[0]);
//             ++inDegree[info[0]];  // 记录入度和图
//         }

//         queue<int> q;
//         for (int i = 0; i < numCourses; i++) {
//             if (inDegree[i] == 0) {
//                 q.push(i);
//             }
//         }

//         int visited = 0;
//         while (!q.empty()) {
//             ++visited;
//             int u = q.front();
//             q.pop();
//             for (int v : graph[u]) {
//                 --inDegree[v];
//                 if (inDegree[v] == 0) {
//                     q.push(v);
//                 }
//             }
//         }
//         return visited == numCourses;
//     }
// };

class Solution {
private:
    vector<vector<int>> edges;
    vector<int> visited;  // 记录状态
    bool valid = true;

public:
    void dfs(int u) {
        visited[u] = 1;  // 在搜索中
        for (int v : edges[u]) {
            if (visited[v] == 0) {  // 没有搜索过
                dfs(v);             // 搜索它
                if (!valid)
                    return;
            } else if (visited[v] == 1) {
                valid = false;  // 搜索中，闭环，代表这个深度路径不可行
                return;
            }
        }
        visited[u] = 2;
    }
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        edges.resize(numCourses);
        visited.resize(numCourses);
        for (const auto& info : prerequisites) {
            edges[info[1]].push_back(info[0]);
        }
        for (int i = 0; i < numCourses && valid; i++) {
            if (!visited[i]) {
                dfs(i);
            }
        }
        return valid;
    }
};

File: leetcode_py/p207.py
from typing import List


# class Solution:
#     def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

#         # 广度优先遍历
#         # 图和入度表
#         graph = [[] for _ in range(numCourses)]
#         in_degree = [0 for _ in range(numCourses)]
#         for i, j in prerequisites:
#             graph[i].append(j)
#             in_degree[j] += 1

#         # 表建立完成
#         # 入度为0的节点入队
#         queue = []
#         for i in range(numCourses):
#             if in_degree[i] == 0:
#                 queue.append(i)

#         # 开始遍历
#         while queue:
#             course = queue.pop(0)
#             for i in graph[course]:
#                 in_degree[i] -= 1
#                 if in_degree[i] == 0:
#                     queue.append(i)

#         # 遍历结束，如果所有节点都被遍历，则返回True
#         return all(in_degree[i] == 0 for i in range(numCourses))

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        # 深度优先遍历
        # 图和邻接表
        graph = [[] for _ in range(numCourses)]
        visited = [0 for _ in range(numCourses)]
        for i, j in prerequisites:
            graph[i].append(j)

        # 表建立完成
        # 邻接表建立完成
        # 遍历所有节点，如果节点没有被访问过，则递归访问其邻接表

        # 开始遍历
        def dfs(i):
            if visited[i] == -1:  # 访问锅，闭环了
                return False
            if visited[i] == 1:  # 访问得是已经修好
                return True
            visited[i] = -1
            for j in graph[i]:
                if not dfs(j):
                    return False
            visited[i] = 1
            return True

        for i in range(numCourses):
            if not dfs(i):
                return False
        return True
